solana/solana-ibc/programs/solana-ibc/src/lib.rs

```diff
@@ -191,6 +201,68 @@ pub mod solana_ibc {
         chain.set_stake((validator).into(), amount)
     }
 
+    /// Sets up new fee collector proposal which wont be changed until the new fee collector
+    /// calls `accept_fee_collector_change`. If the method is called for the first time, the fee
+    /// collector would just be set without needing for any approval.
+    ///
+    pub fn setup_fee_collector<'a, 'info>(
+        ctx: Context<'a, 'a, 'a, 'info, SetupFeeCollector<'info>>,
+        new_fee_collector: Pubkey,
+    ) -> Result<()> {
+        let private_storage = &mut ctx.accounts.storage;
+
+        let signer = ctx.accounts.fee_collector.key();
+
+        if private_storage.fee_collector == Pubkey::default() {
+            private_storage.fee_collector = new_fee_collector;
+        } else if signer == private_storage.fee_collector {
+            private_storage.new_fee_collector_proposal = Some(new_fee_collector)
+        } else {
+            return Err(error::Error::InvalidFeeCollector)
+                .map_err(|err| error!((&err)));
+        }
+
+        Ok(())
+    }
+
+    pub fn accept_fee_collector_change<'a, 'info>(
+        ctx: Context<'a, 'a, 'a, 'info, SetupFeeCollector<'info>>,
+    ) -> Result<()> {
+        let private_storage = &mut ctx.accounts.storage;
+
+        let signer = ctx.accounts.fee_collector.key();
+
+        if let Some(new_admin) = private_storage.new_fee_collector_proposal {
+            if signer != new_admin {
+                return Err(error::Error::InvalidFeeCollector)
+                    .map_err(|err| error!((&err)));
+            }
+            private_storage.fee_collector = new_admin;
+            private_storage.new_fee_collector_proposal = None;
+        } else {
+            return Err(error::Error::FeeCollectorChangeProposalNotSet)
+                .map_err(|err| error!((&err)));
+        }
+
+        Ok(())
+    }
+
+    pub fn collect_fees<'a, 'info>(
+        ctx: Context<'a, 'a, 'a, 'info, CollectFees<'info>>,
+    ) -> Result<()> {
+        let fee_collector_balance = ctx.accounts.fee_account.lamports();
+        let minimum_rent_balance = ctx.accounts.rent.minimum_balance(0);
+
+        let total_fees_collected = fee_collector_balance - minimum_rent_balance;
+
+        **ctx.accounts.fee_account.try_borrow_mut_lamports().unwrap() -=
+            total_fees_collected;
+        **ctx.accounts.fee_collector.try_borrow_mut_lamports().unwrap() +=
+            total_fees_collected;
+
+        Ok(())
+    }
```

Comment:
I don’t understand what this is supposed to do. Why are we transferring all balance from an account? And why are we assuming the account has zero size? (I’m also not sure if we don’t need to use system_instruction::transfer to do the transfer).

Reply:
the fees are collected in a PDA so that the fees can be returned during timeouts and failed transfers. So this method transfers the fee collected back to the fee collector account set.

Reply:
But then i think we should transfer only part of the fees

Reply:
I don’t think it should be that complicated. ‘Multisig’ is perhaps the wrong term to use her, but from what I’m seeing it should be relatively easy to prepare a transaction and than have it signed by multiple accounts via the partialSign or even addSignature. There is coordination cost on the client side but the two accounts have to coordinate anyway in some capacity.

Reply:
well it kinda makes sense.